add unit test to show a package import bug

jb090979 opened the pull request:
Currently a package has references to all imported namespaces. The processor does not consider the import directive correctly. A import directive defines which namespace is visible to all namespaces define in the current file not in the current package. in this example there are 3 fidl files correspond to the same package "P". In P2.fidl Interface "I2" uses a type of P.T1 which is not imported in this file. So an exception  ( T1Enumnot found) should be raised by the processor. Currently there is no exception.

In this unittest the package references point to different objects altough there is just one package.

Fix:
Each namespace has a list with references to all visible namespaces.  So the processor can check if the current namespace is able to access the used types. This a similar behavior like the "C" include.


The code you do not merge (because you was not entirely clear about its purpose) contains the fix for this issue. Hopefully this unittest helps you to understand my purpose.

Commit:
add unit test to show an package import bug

Currently a package has refernces to all imported namespaces. The processor does not cosnider the import directive. A import directive defines which namespace is visible to all namespaces define in the current file not in the current package.

In this unittest the package references point to different objects altough there is just one package.

Fix:
Each namespace has a list of namespace references of all visible namespaces.

jb090979 commented:
If you can confirm this bug I would add the necessary changes to this request.

jb090979 commented:
I had to change some of the other unittest to get it running on my host (Windows with pycharm).

zayfod commented:
I agree with the general statement.

My conclusion is that we should be doing reference resolution for the package in a file, before (potentially) merging it into a package with the same name, already known to the Processor.

Does this make sense?

Can we come up with a few simple test cases for the correct general behavior?

jb090979 commented:
My conclusion is that we should be doing reference resolution for the package in a file, before (potentially) merging it into a package with the same name, already known to the Processor.

**Does this make sense?**

This makes sense if you only think about reference resolution. But after merging a package into another one with the same name you lost the information which imported namespaces are visible in the current namespace. This information is essential for a code generator. My generator generates a hpp/cpp file for each namespace. I have to generate an #include directive for all imported namespaces. With the current solution I only can generate #include directive for all imported namespaces in the whole package. So I inlcude too much headerfiles. I am not very excited about that.

Therfore my approach is that each namespace has a list of imported namespaces and not the package. See my ast definition.

```
class Namespace(object):

    __metaclass__ = ABCMeta

    def __init__(self, name, flags=None, members=None, comments=None):
        self.package = None
        self.name = name
        self.flags = flags if flags else []         # Unused
        self.version = None
        self.typedefs = OrderedDict()
        self.enumerations = OrderedDict()
        self.structs = OrderedDict()
        self.arrays = OrderedDict()
        self.maps = OrderedDict()
        self.constants = OrderedDict()
        self.comments = comments if comments else OrderedDict()
        # list of imported namespaces 
       self.namespace_references = []
```

Does this make sense?
Is there anything wrong in my solution?

**Can we come up with a few simple test cases for the correct general behavior?**

This is a good idea. During implementing my solution I was not sure about allow or not allow transitive imports.  I decide not to support transitive imports. So each fidl file have to import the namespaces it use directly. This behavior is different to C include. Defining unit test is a very good way to specify the import behavior.

zayfod committed:
Rework the processor to use absolute file specifications.

zayfod commented:
OK, I . understand the problem now and agree that the current behavior is not correct.

The current behavior is by design - I have misunderstanding how imports should work according to the Franca spec.

Would you, please make a proposal, based on the current `dev` branch?

zayfod committed:
Add base for Franca IDL compliance testing.

zayfod committed:
Fix Python 3 test compatibility.

jb090979 commented:
I make an proposal. But the next days my time is limited. So it take some time.

Commit:
fix bug in import_file()

if fspec is a relative path it could be that a file is imported twice.

Fix:
First determine absolute path of fidl file and then checks if the file is already imported.

zayfod committed:
Add support for unions.

zayfod committed:
Fixed tool descriptions.

zayfod committed:
Add support for unions in fidl_dump.py and update README.md .

Commit:
fix import bug

- replace import_stin() function with temporary file creation
--> it is not need anymore
- add a list of imported namespace references to each namespace.
- adjust symbol resolver to use namespace_references instead of package_references

Commit:
add missing check

Commit:
improve comments

jb090979 commented:
I create a new request. My initial solution also has mistakes by design. I forgot to consider circular package dependencies.
